operators/pkg/controller/elasticsearch/driver/downscale_test.go

```diff
@@ -0,0 +1,709 @@
+					"sset3Replicas-1": {Name: "sset3Replicas-1"},
+					"sset3Replicas-2": {Name: "sset3Replicas-2"},
+					// nodes from 2nd sset
+					"sset4Replicas-0": {Name: "sset4Replicas-0"},
+					"sset4Replicas-1": {Name: "sset4Replicas-1"},
+					"sset4Replicas-2": {Name: "sset4Replicas-2"},
+					"sset4Replicas-3": {Name: "sset4Replicas-3"},
+				},
+				RoutingTable: esclient.RoutingTable{
+					Indices: map[string]esclient.Shards{
+						"index-1": {
+							Shards: map[string][]esclient.Shard{
+								"0": {
+									// node sset4Replicas-2 cannot leave the cluster because of this shard
+									{Index: "index-1", Shard: 0, State: esclient.STARTED, Node: "sset4Replicas-2"},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		esClient: esClient,
+	}
+
+	// request downscale from 3 to 1 replicas
+	sset3ReplicasDownscaled := *sset3Replicas.DeepCopy()
+	sset3ReplicasDownscaled.Spec.Replicas = common.Int32(1)
+	// request downscale from 4 to 2 replicas
+	sset4ReplicasDownscaled := *sset4Replicas.DeepCopy()
+	sset4ReplicasDownscaled.Spec.Replicas = common.Int32(2)
+	requestedStatefulSets := sset.StatefulSetList{sset3ReplicasDownscaled, sset4ReplicasDownscaled}
+
+	// do the downscale
+	results := HandleDownscale(downscaleCtx, requestedStatefulSets, actualStatefulSets)
+	require.False(t, results.HasError())
+
+	// data migration should have been requested for all nodes leaving the cluster
+	require.True(t, esClient.ExcludeFromShardAllocationCalled)
+	require.Equal(t, "sset3Replicas-2,sset3Replicas-1,sset4Replicas-3,sset4Replicas-2", esClient.ExcludeFromShardAllocationCalledWith)
+
+	// only part of the expected replicas of sset4Replicas should be updated,
+	// since a node still needs to migrate data
+	sset4ReplicasExpectedAfterDownscale := *sset4Replicas.DeepCopy()
+	sset4ReplicasExpectedAfterDownscale.Spec.Replicas = common.Int32(3)
+	expectedAfterDownscale := []appsv1.StatefulSet{sset3ReplicasDownscaled, sset4ReplicasExpectedAfterDownscale}
+
+	// voting config exclusion should have been added for leaving masters
+	require.True(t, esClient.AddVotingConfigExclusionsCalled)
+	require.Equal(t, []string{"sset3Replicas-2", "sset3Replicas-1", "sset4Replicas-3"}, esClient.AddVotingConfigExclusionsCalledWith)
+
+	// compare what has been updated in the apiserver with what we would expect
+	var actual appsv1.StatefulSetList
+	err := k8sClient.List(&client.ListOptions{}, &actual)
+	require.NoError(t, err)
+	require.Equal(t, expectedAfterDownscale, actual.Items)
+
+	// running the downscale again should give the same results
+	results = HandleDownscale(downscaleCtx, requestedStatefulSets, actual.Items)
+	require.False(t, results.HasError())
+	err = k8sClient.List(&client.ListOptions{}, &actual)
+	require.NoError(t, err)
+	require.Equal(t, expectedAfterDownscale, actual.Items)
+
+	// once data migration is over the complete downscale should go through
+	downscaleCtx.observedState.ClusterState.RoutingTable = esclient.RoutingTable{}
```

Comment:
nit: I guess that if you want something more realistic the routing table should not be empty. The shard 0 of `index-1` should have moved to an other node?